create feature/drawing_adapter

vitali-anku opened the pull request:
- create adapter
- create holder for adapter
- create favorite

CraggyHaggy commented on app/src/main/java/com/testtask/apptesttask/entity/charactrers/Character.kt:
```diff
@@ -1,18 +1,6 @@
-    @SerializedName("stories") val stories: StoryList,
-    @SerializedName("events") val events: EventList,
-    @SerializedName("series") val series: SeriesList
+    val apiCharacter: ApiCharacter,
```
Don't pass Data entity into Domain, but just remove `@SerializedName` and return all fields and fill with server values.

CraggyHaggy commented on app/src/main/java/com/testtask/apptesttask/model/repository/CharactersRepository.kt:
```diff
@@ -1,20 +1,44 @@
\ No newline at end of file
+
+    fun favoriteCharacter(character: Character, favorite: Boolean) {
+        character.favorite = !favorite
```
Make this method return `Single<Character>`, also you must save into `Prefs` this value.

CraggyHaggy commented on app/src/main/java/com/testtask/apptesttask/entity/charactrers/Character.kt:
```diff
@@ -1,18 +1,6 @@
-    @SerializedName("events") val events: EventList,
-    @SerializedName("series") val series: SeriesList
+    val apiCharacter: ApiCharacter,
+    var favorite: Boolean
```
Make it `val` and use `copy` method instead of mutability.

CraggyHaggy commented on app/src/main/java/com/testtask/apptesttask/presentation/characters/CharactersPresenter.kt:
```diff
@@ -22,9 +24,15 @@ class CharactersPresenter @Inject constructor(
                 .doAfterTerminate { viewState.hideProgress() }
                 .subscribe(
-                    { viewState.showCharacters(it.data.results) },
+                    {
```
Make one line.

CraggyHaggy commented on app/src/main/java/com/testtask/apptesttask/model/interactor/characters/CharactersInteractor.kt:
```diff
@@ -10,7 +10,10 @@ class CharactersInteractor @Inject constructor(
+    fun setFavoriteCharacter(character: Character, favorite: Boolean): Completable =
+            Completable.fromAction {
+                charactersRepository.favoriteCharacter(character, favorite)
+            }
```
You make it on `UI-thread`.

CraggyHaggy commented on app/src/main/java/com/testtask/apptesttask/model/repository/CharactersRepository.kt:
```diff
@@ -1,20 +1,44 @@
             .getCharacters(appConfig.ts, appConfig.publicKey, appConfig.hash)
+            .flatMap { characterDataWrapper -> Single.just(characterDataWrapper.data.results) }
+
+    fun getCharacters(): Single<List<Character>> = Single
```
Move `Single` to new line.

CraggyHaggy commented on app/src/main/java/com/testtask/apptesttask/model/repository/CharactersRepository.kt:
```diff
@@ -1,20 +1,44 @@
+                Single.just(sharedPreference.favoritesCharacters),
+                BiFunction<List<ApiCharacter>, List<Boolean>?, List<Character>> { apiCharacters, favorites ->
+                    apiCharacters.map { apiCharacter ->
+                        val position = apiCharacters.indexOf(apiCharacter)
```
Refactor this method according to `Map` impl.

CraggyHaggy commented on app/src/main/java/com/testtask/apptesttask/model/storage/Prefs.kt:
```diff
@@ -0,0 +1,32 @@
+    private val sharedPreferences =
+            context.getSharedPreferences(PREF_TAG, Context.MODE_PRIVATE)
+
+    var favoritesCharacters: MutableList<Boolean>?
```
Instead of this implementation, keep ` Map<Id, Character>` with only favorite objects.

CraggyHaggy commented on app/src/main/java/com/testtask/apptesttask/model/storage/Prefs.kt:
```diff
@@ -0,0 +1,32 @@
+class Prefs @Inject constructor(context: Context, val gson: Gson) {
+
+    private val PREF_TAG = "characters"
+    private val FAVORITE_CHARACTER = "favorite_characters"
```
Rename this as `FAVORITE_CHARACTERS`.

CraggyHaggy commented on app/src/main/java/com/testtask/apptesttask/model/storage/Prefs.kt:
```diff
@@ -0,0 +1,32 @@
+import com.google.gson.Gson
+import javax.inject.Inject
+
+class Prefs @Inject constructor(context: Context, val gson: Gson) {
```
Format this constructor.

CraggyHaggy commented on app/src/main/java/com/testtask/apptesttask/model/repository/CharactersRepository.kt:
```diff
@@ -1,20 +1,44 @@
     private val appConfig: AppConfig,
-    private val schedulers: SchedulersProvider
+    private val schedulers: SchedulersProvider,
+    private val sharedPreference: Prefs
```
Rename field as `prefs`.

CraggyHaggy commented on app/src/main/java/com/testtask/apptesttask/presentation/characters/CharactersPresenter.kt:
```diff
@@ -22,9 +24,15 @@ class CharactersPresenter @Inject constructor(
                 .connect()
     }
+
+    fun favoriteCharacter(character: Character, favorite: Boolean): Completable {
```
`Presenter` shouldn't return anything.

CraggyHaggy replied:
Pass `position` in the adapter to update it later.

CraggyHaggy commented on app/src/main/java/com/testtask/apptesttask/ui/characters/CharactersFragment.kt:
```diff
@@ -38,18 +40,22 @@ class CharactersFragment : BaseFragment(), CharactersView {
 
     override fun showCharacters(characters: List<Character>) {
-        //TODO characters fragment (Add implementation this method).
+        adapter = CharactersAdapter(context!!, characters, this)
```
You must create `Adapter` in `onViewCreated` and later just fill it.

CraggyHaggy commented on app/src/main/java/com/testtask/apptesttask/ui/global/CharactersAdapter.kt:
```diff
@@ -0,0 +1,40 @@
+
+class CharactersAdapter constructor(
+    val context: Context,
+    val characters: List<Character>,
```
Remove `characters` param, and make `context` as `private`.

CraggyHaggy commented on app/src/main/java/com/testtask/apptesttask/ui/global/CharactersAdapter.kt:
```diff
@@ -0,0 +1,40 @@
+class CharactersAdapter constructor(
+    val context: Context,
+    val characters: List<Character>,
+    private val listener: CharactersView
```
Don't pass `CharactersView`, instead pass lambda as `(int, Character) -> Unit`.

CraggyHaggy commented on app/src/main/java/com/testtask/apptesttask/ui/global/CharactersAdapter.kt:
```diff
@@ -0,0 +1,40 @@
+        val itemView = LayoutInflater.from(context).inflate(null, view, false)
+        val holder = CharactersHolder(itemView)
+
+        holder.itemView.setOnClickListener {
```
Use `DiffUtil` to notify `RecyclerView`.

vitali-anku commented:
- add reight entity Character
- add Diffutil
- edited adapter, fragment, presenter, prefs and repositpory and interactor

vitali-anku commented:
- edited Presenter (copy class)
- edited var by Character field
- edited save prefs in Repository

CraggyHaggy commented on app/src/main/java/com/testtask/apptesttask/model/interactor/characters/CharactersInteractor.kt:
```diff
@@ -2,18 +2,21 @@ package com.testtask.apptesttask.model.interactor.characters
 
-    fun setFavoriteCharacter(character: Character, favorite: Boolean): Completable =
+    fun setFavoriteCharacter(id: Int, character: Character): Completable =
             Completable.fromAction {
```
Make `Repository` returns `Completable` instead of `Interactor` composing it.

CraggyHaggy commented on app/src/main/java/com/testtask/apptesttask/model/interactor/characters/CharactersInteractor.kt:
```diff
@@ -2,18 +2,21 @@ package com.testtask.apptesttask.model.interactor.characters
-                charactersRepository.favoriteCharacter(character, favorite)
+                charactersRepository.favorCharacter(id, character)
             }
+                    .observeOn(schedulers.ui())
```
Move thread logic to `Repository`.

CraggyHaggy commented on app/src/main/java/com/testtask/apptesttask/entity/charactrers/Character.kt:
```diff
@@ -1,6 +1,18 @@
+    val stories: StoryList,
+    val events: EventList,
+    val series: SeriesList,
+    var favorite: Boolean?
```
Why it is `nullable`? Refactor it, please.

CraggyHaggy commented on app/src/main/java/com/testtask/apptesttask/model/repository/CharactersRepository.kt:
```diff
@@ -14,31 +14,55 @@ class CharactersRepository @Inject constructor(
+    private fun getPrefs(): Single<Map<Int, Character>> =
+            Single.just(prefs.favoritesCharacters)
+
+    fun getCharacters(): Single<List<Character>> =
```
Change method order according to call:
```
fun a() {
       b()
       c()
}

private b()
private c()
```

CraggyHaggy commented on app/src/main/java/com/testtask/apptesttask/model/repository/CharactersRepository.kt:
```diff
@@ -14,31 +14,55 @@ class CharactersRepository @Inject constructor(
 ) {
 
-    private fun getApiCharacters(): Single<List<ApiCharacter>> = marvelService
+    private fun getApiCharacters() = marvelService
```
Remove this method and use `marvelService.getCharacters` in `Single.zip` with `characterDataWrapper` in `BiFunction`.

CraggyHaggy commented on app/src/main/java/com/testtask/apptesttask/model/repository/CharactersRepository.kt:
```diff
@@ -14,31 +14,55 @@ class CharactersRepository @Inject constructor(
 
-    fun getCharacters(): Single<List<Character>> = Single
-            .zip(
+    private fun getPrefs(): Single<Map<Int, Character>> =
```
Remove this method and use `Single.just(prefs.favoritesCharacters)` in `Single.zip`.

CraggyHaggy commented on app/src/main/java/com/testtask/apptesttask/model/repository/CharactersRepository.kt:
```diff
@@ -14,31 +14,55 @@ class CharactersRepository @Inject constructor(
+                            it.stories,
+                            it.events,
+                            it.series,
+                            if (favorites.containsKey(id)) {
```
Evaluating favorite state will be crashed according to `Map<Int, Character>?`. Refactor this logic, please.
Make this logic one line.

CraggyHaggy commented on app/src/main/java/com/testtask/apptesttask/model/repository/CharactersRepository.kt:
```diff
@@ -14,31 +14,55 @@ class CharactersRepository @Inject constructor(
-    fun favoriteCharacter(character: Character, favorite: Boolean) {
-        character.favorite = !favorite
+    fun favorCharacter(id: Int, character: Character) {
+        prefs.favoritesCharacters?.let {
```
Why do you call it on `UI-thread`?

CraggyHaggy commented on app/src/main/java/com/testtask/apptesttask/model/storage/Prefs.kt:
```diff
@@ -2,31 +2,35 @@ package com.testtask.apptesttask.model.storage
 
-class Prefs @Inject constructor(context: Context, val gson: Gson) {
+class Prefs @Inject constructor(
+    context: Context,
```
Make it `private val`.

CraggyHaggy commented on app/src/main/java/com/testtask/apptesttask/model/storage/Prefs.kt:
```diff
@@ -2,31 +2,35 @@ package com.testtask.apptesttask.model.storage
 
-    private fun convertArrayToMutableList(valueString: String?): MutableList<Boolean> {
-        val mutableList = mutableListOf<Boolean>()
+    private fun convertArrayToMutableList(valueString: String?): MutableMap<Int, Character> {
```
Remove this logic and keep `Map` as `JSON` -> https://stackoverflow.com/questions/8360836/gson-is-there-an-easier-way-to-serialize-a-map.

CraggyHaggy commented on app/src/main/java/com/testtask/apptesttask/presentation/characters/CharactersPresenter.kt:
```diff
@@ -25,14 +26,23 @@ class CharactersPresenter @Inject constructor(
-    fun favoriteCharacter(character: Character, favorite: Boolean): Completable {
-        return charactersInteractor.setFavoriteCharacter(character, favorite)
+    fun favoritCharacter(position: Int) {
+        val character = characters[position]
```
Make this RxJava chain method.

CraggyHaggy commented on app/src/main/java/com/testtask/apptesttask/presentation/characters/CharactersView.kt:
```diff
@@ -12,6 +12,5 @@ interface CharactersView : MvpView {
     fun hideProgress()
-
-    fun favorCharacter(character: Character)
+    fun favorCharacter(characters: List<Character>)
```
Remove this method and use `showCharacters` always.

CraggyHaggy commented on app/src/main/java/com/testtask/apptesttask/ui/characters/CharactersFragment.kt:
```diff
@@ -44,15 +46,15 @@ class CharactersFragment : BaseFragment(), CharactersView {
 
     override fun showError(message: String) {
-        //TODO apiCharacters fragment (Add implementation this method).
+        Log.e("Error", message)
```
If you add logs to vcs, it should be displayed only in debug.

CraggyHaggy commented on app/src/main/java/com/testtask/apptesttask/ui/global/CharactersAdapter.kt:
```diff
@@ -1,28 +1,24 @@
     val context: Context,
-    val characters: List<Character>,
-    private val listener: CharactersView
+    private val listener: (position: Int) -> Unit
```
Rename listener according to context.

CraggyHaggy commented on app/src/main/java/com/testtask/apptesttask/ui/global/CharactersAdapter.kt:
```diff
@@ -34,7 +30,57 @@ class CharactersAdapter constructor(
+        position: Int,
+        payloads: MutableList<Any>
+    ) {
+        super.onBindViewHolder(holder, position, payloads)
```
Do you need `super` call?

CraggyHaggy commented on app/src/main/java/com/testtask/apptesttask/ui/global/CharactersAdapter.kt:
```diff
@@ -34,7 +30,57 @@ class CharactersAdapter constructor(
+    ) {
+        super.onBindViewHolder(holder, position, payloads)
+
+        val character = characters[position]
```
You don't need two versions of `onBindViewHolder`. If `payloads` are empty, then use `holder.bind(character)`, else `holder.updateStar(character.favorite!!)`.

CraggyHaggy commented on app/src/main/java/com/testtask/apptesttask/ui/global/CharactersAdapter.kt:
```diff
@@ -34,7 +30,57 @@ class CharactersAdapter constructor(
+        holder.updateStar(character.favorite!!)
+    }
+
+    fun updateCharacters(newCharacters: List<Character>) {
```
Rename this method as `setCharacters(characters:..)`.

CraggyHaggy commented on app/src/main/java/com/testtask/apptesttask/ui/global/CharactersAdapter.kt:
```diff
@@ -1,28 +1,24 @@
+    private val listener: (position: Int) -> Unit
 ) : RecyclerView.Adapter<CharactersHolder>() {
 
+    private var characters = mutableListOf<Character>()
```
Rename field as `items`.

vitali-anku commented:
- edited naming methods and fields
- deleted some methods
- edited logic repository/presenter

vitali-anku commented:
- edited URI in MarvelService
- edited CharactersRepository
- removed nullable preferences value

vitali-anku commented:
Sorry, I made a little mud

vitali-anku commented:
- removed textView in fragment_characters
- format menu menu file

CraggyHaggy commented on app/src/main/java/com/testtask/apptesttask/model/interactor/characters/CharactersInteractor.kt:
```diff
@@ -2,21 +2,15 @@ package com.testtask.apptesttask.model.interactor.characters
-                charactersRepository.favorCharacter(id, character)
-            }
-                    .observeOn(schedulers.ui())
+    fun setFavoriteCharacter(position: Int): Single<MutableList<Character>>? =
```
Return `Single<Character` instead of `List<Character`.

CraggyHaggy commented on app/src/main/java/com/testtask/apptesttask/model/repository/CharactersRepository.kt:
```diff
@@ -17,50 +17,55 @@ class CharactersRepository @Inject constructor(
-
-    private fun getPrefs(): Single<Map<Int, Character>> =
-            Single.just(prefs.favoritesCharacters)
+    private val characters = mutableListOf<Character>()
```
Remove this collection, because we don't implement cache at this time.

CraggyHaggy commented on app/src/main/java/com/testtask/apptesttask/model/repository/CharactersRepository.kt:
```diff
@@ -17,50 +17,55 @@ class CharactersRepository @Inject constructor(
-                getPrefs(),
+                marvelService
+                        .getCharacters(appConfig.ts, appConfig.publicKey, appConfig.hash)
+                        .flatMap { characterDataWrapper ->
```
Use `map` instead of `flatMap`.

CraggyHaggy commented on app/src/main/java/com/testtask/apptesttask/model/repository/CharactersRepository.kt:
```diff
@@ -17,50 +17,55 @@ class CharactersRepository @Inject constructor(
 
-    fun favorCharacter(id: Int, character: Character) {
+    fun favorCharacter(position: Int): Single<MutableList<Character>>? {
+        val character = characters[position].copy(favorite = !characters[position].favorite)
```
You implement all this logic on UI-thread and just pass it into the Rx-chain. Save these values into `Prefs`.

CraggyHaggy commented on app/src/main/java/com/testtask/apptesttask/model/storage/Prefs.kt:
```diff
@@ -18,19 +18,19 @@ class Prefs @Inject constructor(
         }
 
-    private fun convertArrayToMutableList(valueString: String?): MutableMap<Int, Character> {
+    private fun convertCharacters(gsonString: String?): MutableMap<Int, Character> {
```
Rename `gsonString` as `json`.

CraggyHaggy commented on app/src/main/java/com/testtask/apptesttask/model/repository/CharactersRepository.kt:
```diff
@@ -17,50 +17,55 @@ class CharactersRepository @Inject constructor(
-                                false
+                        characters.add(
+                            id, Character(
+                                id,
```
Use `with(it)` construction.

CraggyHaggy commented on app/src/main/java/com/testtask/apptesttask/model/repository/CharactersRepository.kt:
```diff
@@ -28,44 +28,50 @@ class CharactersRepository @Inject constructor(
                         },
                 Single.just(prefs.favoritesCharacters),
                 BiFunction<List<ApiCharacter>, Map<Int, Character>?, List<Character>> { apiCharacters, favorites ->
+                    apiCharacters.size
```
Remove these two lines.

vitali-anku commented:
- edited parameters and return value in CharactersInteractor
- replaced flatMap on map in Repository
- edited Char-presenter
- edited name parameter in Prefs-class

CraggyHaggy commented on app/src/main/java/com/testtask/apptesttask/model/interactor/characters/CharactersInteractor.kt:
```diff
@@ -11,6 +11,6 @@ class CharactersInteractor @Inject constructor(
 
-    fun setFavoriteCharacter(position: Int): Single<MutableList<Character>>? =
-            charactersRepository.favorCharacter(position)
+    fun setFavoriteCharacter(character: Character): Single<Character> =
```
Remove type definition and rename method as `favorCharacter`.

CraggyHaggy commented on app/src/main/java/com/testtask/apptesttask/model/repository/CharactersRepository.kt:
```diff
@@ -17,22 +17,17 @@ class CharactersRepository @Inject constructor(
-                BiFunction<List<ApiCharacter>, Map<Int, Character>?, List<Character>> { apiCharacters, favorites ->
-                    apiCharacters.size
-                    favorites.size
+                BiFunction<List<ApiCharacter>, Map<Int, Character>?,
```
`Prefs` should return not nullable return.

CraggyHaggy commented on app/src/main/java/com/testtask/apptesttask/model/repository/CharactersRepository.kt:
```diff
@@ -46,32 +41,29 @@ class CharactersRepository @Inject constructor(
-            favoriteCharacters[id] = character
-            prefs.favoritesCharacters = favoriteCharacters
-        }
+            Single.just(character)
```
Change with `Single.fromCallable()`.

CraggyHaggy commented on app/src/main/java/com/testtask/apptesttask/model/repository/CharactersRepository.kt:
```diff
@@ -46,32 +41,29 @@ class CharactersRepository @Inject constructor(
+                        }
+                        favoriteCharacter
+                    }
+                    .subscribeOn(schedulers.newThread())
```
Why do you use `newThread()` instead of `io()`?

CraggyHaggy commented on app/src/main/java/com/testtask/apptesttask/model/repository/CharactersRepository.kt:
```diff
@@ -46,32 +41,29 @@ class CharactersRepository @Inject constructor(
+            Single.just(character)
+                    .map {
+                        val favoriteCharacter = it.copy(favorite = !it.favorite)
+                        val favoriteCharacters = prefs.favoritesCharacters
```
Rename `favoriteCharacters` as `characters` and `favoriteCharacter` as `newCharacter`.

CraggyHaggy commented on app/src/main/java/com/testtask/apptesttask/model/repository/CharactersRepository.kt:
```diff
@@ -46,32 +41,29 @@ class CharactersRepository @Inject constructor(
-    }
+                        if (favoriteCharacters.contains(it.id)) {
+                            favoriteCharacters.remove(it.id)
+                            prefs.favoritesCharacters = favoriteCharacters
```
This line is duplicated in two branches. Move it from the condition.

CraggyHaggy commented on app/src/main/java/com/testtask/apptesttask/model/storage/Prefs.kt:
```diff
@@ -25,9 +25,9 @@ class Prefs @Inject constructor(
 
-    private fun convertCharacters(gsonString: String?): MutableMap<Int, Character> {
+    private fun convertCharacters(json: String?): MutableMap<Int, Character> {
         val mutableList = mutableMapOf<Int, Character>()
```
Rename it as `characters`.

vitali-anku commented:
- removed type definition and removed method in interactor
- removed nullable and edited method favorCharacter Prefs in Repository
- rename smt in CharPresenter and Prefs
- update version kotlin

CraggyHaggy commented on app/src/main/java/com/testtask/apptesttask/model/interactor/characters/CharactersInteractor.kt:
```diff
@@ -2,15 +2,15 @@ package com.testtask.apptesttask.model.interactor.characters
 
-    fun getCharacters(): Single<List<Character>> = charactersRepository.getCharacters()
+    fun getCharacters() =
+            charactersRepository.getCharacters()
```
Make it one line method.

vitali-anku commented:
formated CharactersRepository

vitali-anku commented:
- formated CharactersInteractor
